### ISSUE TITLE
Chinese map editing

### ISSUE THREAD

**Issue by fabianlam** [olikraus/u8g2#710]
Hello again! I would like to have your helping hand. If I would like to add few Chinese words into the map file. Is there anything or instruction to do? Thank you!

**Comment by olikraus** [comment]
What exactly is your goal? Do you want to create your u8g2 font with a special unicode subset?
Or do you want to extend and update the existing map files for u8g2?

This is a list of all map files currently in use for Chinese language:
https://github.com/olikraus/u8g2/blob/master/tools/font/build/chinese1.map
https://github.com/olikraus/u8g2/blob/master/tools/font/build/chinese2.map
https://github.com/olikraus/u8g2/blob/master/tools/font/build/chinese3.map
https://github.com/olikraus/u8g2/blob/master/tools/font/build/gb2312.map
https://github.com/olikraus/u8g2/blob/master/tools/font/build/gb2312a.map
https://github.com/olikraus/u8g2/blob/master/tools/font/build/gb2312b.map

You can create a pull request if you see a need to update them.

**Comment by RobotGB** [comment]
You would need a list of the characters needed. The format is
$<hexadecimal unicode of character>;
e.g. "$4E00;" for 一 (character for 'one').

**Comment by fabianlam** [comment]
Thank you and let me try again

**Comment by fabianlam** [comment]
Yeah, as the Chinese font inside the database may not have he compete glyph. I would like to add some inside. For e.g. shalll I just add the Chinese font Hexdecimal into the map is ok?

**Comment by olikraus** [comment]
Hmm... I still do not understand what you want to do. 
Can you give an example? Like which glyph is missing in which font? Then I can probably better explain what you need to do.
Github issue tracker here should accept Chinese letters, so you could use a Chinese glyph also here.

**Comment by fabianlam** [comment]
Sorry for my bad wording. Chinese font have two separated division. One is Simplified Chinese and the Second is traditional Chinese. The Chinese 1,2 &3 . Map all are simplified Chinese font. I actually want to convert the Microsoft MingLiu into the font set and used with the U8G2. Could you plz help and let me how to customise my own font set? Appreciated.

**Comment by olikraus** [comment]
ok, let me try to guess what you want:
You have a .ttf font (MingLiu) .
You want a u8g2 font out of it, right?

Steps for the creation of a u8g2 from a TTF file:

Process: .ttf --> .bdf --> .c (u8g2 font c code)
For the .ttf to .bdf conversion you need an external tool like otf2bdf or fontforge.

For the .bdf to .c conversion, you need bdfconv.exe from the u8g2 project, which is available for download here:
https://github.com/olikraus/u8g2/tree/master/tools/font/bdfconv
Instructions for the bdf to .c conversion are also mentioned here:

https://github.com/olikraus/u8g2/blob/master/doc/faq.txt#L203

Now there is another problem: In most cases you do not want all the characters from the original .ttf file in your embedded device. Instead usually a subset is required.
This information has to be given to "bdfconv.exe". By default bdfconv.exe will use the uncodes 32 to 255 only, which anyways is not sufficient. This means: You MUST tell bdfconv.exe which characters to copy from the .bdf file to the .c font format.

There are two ways todo this: Directly via commandline (-m option) or by applying an external file (-M option). In case of Chinese languages the external file might be more suitable. So you probably need a file (called "map file" due to historical reasons)  which contains all the chars you need in the resulting u8g2 font.

The the task for you is to put together a list of unicode numbers, which make up your list of traditional glyphs. You have to format these numbers and put it into the map file as mentioned by @RobotGB .
Then only you can start bdfconv.exe with the provided list of glyphs.

I personally can not provide this list (because of lacking knowledge on Chinese language and because I do not know your constrains on your project). You can only look at the above linked map files as an example.
If these files do not fit, you need to create one.

A quick google search revealed this:
https://chinese.stackexchange.com/questions/22484/list-of-all-traditional-chinese-characters

So you probably need to convert one of the lists there into a file, which follows the bdfconv.exe syntax.

**Comment by olikraus** [comment]
Let me know if you have further questions.:)

**Comment by fabianlam** [comment]
Sure, thank you.

**Comment by fabianlam** [comment]
By the way, do you have any plan. And expand the project with SD Card ? The flexibility will be larger to have a big database with Asian fonts. Just a suggestion 🙏

**Comment by olikraus** [comment]
Feel free to submit pull requests for this ;-)

My personal opinion: If you need a SD Card for storing fonts, then you might have the wrong controller. Still the glyphs do not eat up that much of Flash ROM. Maybe 300KB. Do you really want to use a SD card, which is capable of storing 100000 times more data?

**Comment by AnHardt** [comment]
Hmm. 300kB is about a single side single density 5 1/4'' floppy. At the beginning of my career (VIC20) i thought - this giant beast will never get filled. ;-)

**Comment by olikraus** [comment]
My first C64 programs (10kB or so) had to fit on tape. It took minutes to load and save that basic program. Still in that 64KB of RAM for a commodore C64 people had created amazing tools.

**Comment by fabianlam** [comment]
Hello. May I have one more stupid question ? I have convert the MingLiu.ttf into the MingLiu.bdf by the Guitools. Then used the Bdfconv and extract the list from the Chinese1.map and then to C file. But shall I copy and place the MingLiu  C file into any directory? As I have complied the Ino file and nothing comes out except the numerical characters. Appreciated for your helping hand again.

**Comment by olikraus** [comment]
Just for crosschecking, what add been the exact options you provided to bdfconv.exe? Maybe you can repeat the full command for bdfconv.exe here.

> But shall I copy and place the MingLiu C file into any directory?

You can just copy and paste the resulting c code into your .ino file.

> As I have complied the Ino file and nothing comes out except the numerical characters.

ok, for further investigation, I need the above mentioned full commandline. You could also attach the .bdf (as zip) for further analysis. Also provide the c code you have used for testing.

**Comment by AnHardt** [comment]
A relative common way is to rename the font.c file to font.h and include it into the main file.

**Comment by fabianlam** [comment]
Thank you everyone’s support. I have successfully to display the Chinese Traditional MingLiu in the SSD1322 display.  I have pasted the code into the INO directly and it works. But dunno why the #include xxxz.h doesn’t work in this scenario. But I think everyone had helped me a lot and really appreciated for instant reply to my question. Have a nice day and thank you again!

**Comment by olikraus** [comment]
Please do not put much effort into the SD CARD problem. I am still not convinced, that it makes sense for u8g2. And one more thing: I rember someone working on this topic already. So maybe search in the issues for SD CARD support.

**Comment by RobotGB** [comment]
By the way @fabianlam I was also trying to include support for an SD card (issue #461). I don't think that it would be fast at all (SD cards read at about 250 KB/sec, but SPI flash is much faster (probably limited only by the processor speed).
However, if you are trying to include support (maybe if you don't need to display different text often), then I might see why you would need it.
Either way, as oliver suggested on issue #461, the best option would be to get a microcontroller with more flash memory (as I have done).
But if you really want support on an SD card, I am also working on one (because I will be using multiple large fonts), but I am almost certain that it will not have the same speed that flash memory offers.

**Comment by RobotGB** [comment]
I've generated a c file with the glyphs from "cj5-13053) [here](https://chinese.stackexchange.com/questions/22484/list-of-all-traditional-chinese-characters) along with a copy of the tga preview file and the map in this compressed file
[u8g2_unifont_tradchinese.zip](https://github.com/olikraus/u8g2/files/2469934/u8g2_unifont_tradchinese.zip)

**Comment by RobotGB** [comment]
Update: this is probably quite late, but @fabianlam I have added support for storing fonts on SD card on my library [U8G2_for_Adafruit_GFX_extmem](https://github.com/RobotGB/U8G2_for_Adafruit_GFX_extmem)
